### ISSUE TITLE
help request for points2nearest cell

### ISSUE THREAD

**Issue by chrishaak** [Pakillo/rSDM#3]
Hi Pakillo,

I'm using this (super-useful-thank you!) function for some SDM work but have run into an issue/potential improvement that I am hoping you may be able to help me sort out.  

I have a spatialpoints object with N points.  Each of these points/locations needs to reference a different layer in a raster stack or brick (the different layers are rasters of the same variable but at different times (i.e.,months), and they have NAs in different places- due to, for example, changes in cloud cover).   So moving points to the nearest non-NA cell, must be layer-specific.  However, it appears that the "layer" argument takes only a single layer as input.

I'm wondering if you could adapt/modify the function to permit it to take a vector (with a length equal to the number of points in "locs") for the "layer" argument, so that each location (each point in the spatialpoints object) is moved based on a distinct raster layer in the brick/stack (specified by layer name)?

In other words, imagine my spatialpoints object (locs) has n=100 coordinates.  These points correspond to different days of the year.  Then I also have a raster stack with, say, 365 layers, one for each day of the year.  Imagine these layers are names as character strings ("day1","day2"......).   

I'd like to be able to run points2nearestcell on the 100 points, but instead of specifying a single raster layer, I'd provide "layer" with a vector of 100 layer names, and the function would relocate each point based on it's corresponding layer name in the vector?

This would be really helpful when dealing with any form of time-varying data where NA values do not always fall in the same locations over time.  

Thanks in advance for your time!!
C

**Comment by Pakillo** [comment]
Hi Chris,

I'm glad you found this useful, and thanks for your suggestion. It looks like it should be doable. I don't have much time these days, but let me have a look to see if I can implement it more or less quickly. I also suspect there may be some solution out there using `sf`...

Cheers

**Comment by Pakillo** [comment]
Hi @chrishaak 

After some thought, I think the best option to do what you want right now is using purrr::pmap to run points2nearestcell over a list of spatial points AND a list of raster layers.

I made a reproducible example to illustrate the workflow::

``` r
library(raster)
#> Loading required package: sp
library(rSDM)

## Create simulated raster stack (2 layers)
r <- raster()
values(r) <- runif(180*360)
ras <- stack(r, r)
ras
#> class      : RasterStack 
#> dimensions : 180, 360, 64800, 2  (nrow, ncol, ncell, nlayers)
#> resolution : 1, 1  (x, y)
#> extent     : -180, 180, -90, 90  (xmin, xmax, ymin, ymax)
#> crs        : +proj=longlat +datum=WGS84 +no_defs 
#> names      :      layer.1,      layer.2 
#> min values : 2.215058e-05, 2.215058e-05 
#> max values :    0.9999855,    0.9999855

## Now remove values (i.e. set NA) some region of each rasterlayer
ras[[1]][, 1:180] <- NA
ras[[2]][1:90, ] <- NA
plot(ras)
```

![](https://i.imgur.com/cn5UUdk.png)

``` r


## Now simulate 2 SpatialPointsDataFrame with points out of the raster layers
pt1 <- locs2sp(data.frame(x = -100, y = 20))
pt2 <- locs2sp(data.frame(x = 50, y = 80))

## pt1 is outside layer 1
occmap(pt1, ras[[1]], psize = 3, main = 'layer = 1')
#> Warning in proj4string(locs): CRS object has comment, which is lost in output
```

![](https://i.imgur.com/adKe4Y4.png)

``` r

## pt2 is outside layer 2
occmap(pt2, ras[[2]], psize = 3, main = 'layer = 2')
#> Warning in proj4string(locs): CRS object has comment, which is lost in output
```

![](https://i.imgur.com/piEmf8Q.png)

``` r


## Using purrr::pmap to run points2nearestcell on multiple raster layers

# Create named list with locations (spatial objects) and layers to which they should be compared
datalist <- list(locs = list(pt1, pt2),
                 layer = list(1, 2))

move <- purrr::pmap(.l = datalist, .f = points2nearestcell, ras = ras, showmap = FALSE)
#>   longitude latitude longitude_new latitude_new distances moved
#> 1      -100       20           0.5         20.5  10281616  TRUE
#> 1 out of 1 points have been moved.
#>   longitude latitude longitude_new latitude_new distances moved
#> 1        50       80          49.5         -0.5   8940470  TRUE
#> 1 out of 1 points have been moved.

## See result
## (move is a list containing the revised SpatialPointsDataFrames)

occmap(move[[1]], ras[[1]], psize = 3, main = 'layer = 1')
#> Warning in proj4string(locs): CRS object has comment, which is lost in output
```

![](https://i.imgur.com/QQ0mm29.png)

``` r
occmap(move[[2]], ras[[2]], psize = 3, main = 'layer = 2')
#> Warning in proj4string(locs): CRS object has comment, which is lost in output
```

![](https://i.imgur.com/Wd9ykod.png)

<sup>Created on 2020-10-10 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>


Hope this helps. Please let me know how it goes!

Cheers
